### DISCUSSION TITLE
How to get reference to python subclass object wrapped as &PyAny

### DISCUSSION THREAD

**Discussion by shivamka1** [PyO3/pyo3#2967] [Questions]
I am trying to use a "serde/serde-json" rust library to serialize and deserialize object to json and vice-versa. Areally simple implementation of which in rust would look something like this:
```rust
#[derive(Serialize, Deserialize, Debug, Clone)]
struct Point {
    x: i32,
    y: i32,
}

#[pyfunction]
pub fn point_serde() {
    let point = Point { x: 1, y: 2 };

    let serialized = serde_json::to_string(&point).unwrap();
    println!("serialized = {}", serialized);

    let deserialized: Point = serde_json::from_str(&serialized).unwrap();
    println!("deserialized = {:?}", deserialized);
}
```

What I am trying to do is allow python users to create say `Point` objects in python and have it serialized and deserialized in rust like so:
```rust
#[derive(Serialize, Deserialize, Debug)]
#[pyclass(subclass)]
pub struct General;

#[pymethods]
impl General {
    #[new]
    fn new() -> Self {
        Self
    }
}

#[pyfunction]
pub fn point_serde(point: &PyCell<General>) -> PyResult<()> {
    let r: &PyAny = point.as_ref();
    println!("{:?}", point.as_ref());

    // let serialized = serde_json::to_string(r).unwrap();
    // println!("serialized = {}", serialized);

    // let deserialized: Point = serde_json::from_str(&serialized).unwrap();
    // println!("deserialized = {:?}", deserialized);

    Ok(())
}
```

```python
class Point(pyo3_example.General):
    def __new__(cls, x, y):
        return super(Point, cls).__new__(cls)

    def __init__(self, x, y) -> None:
        self.x = x
        self.y = y

pyo3_example.point_serde(Point(3, 2))
```
This results in `Point(3,2)`.

What I have done here is that I have create a `General` struct as `pyclass` that derives `Serialize` and `Deserialize` implementations from serde library as a superclass in Rust and `Point` subclass in python. The intention is to allow any class that extends `General` have the capability of being serialized and deserialized in rust. Then I have created a `pyfunction` called `point_serde` that is supposed to do the actual serialization and deserialization. 

When I try to print `point.as_ref()`, it returns `Point(3,2)` correctly. But the type is `&PyAny`. I am unsure how to pass `Point` to `serde_json::to_string(r)` as `r`. Please suggest.

**Comment by adamreichold** [comment]
This will not work: Serde works by syntactically enumerating the fields at compile time and hence it cannot know about the dynamically created Python properties. I think using a Python-oriented JSON serializer like Python's built-in `json` module would be preferable.
